open-craft/opencraft
# (SE-125) Sandbox billing: basic report

TASK: [SE-125](https://tasks.opencraft.com/browse/SE-125)

### Requirements
In this PR you need to test the following:

Description | Developed | Tested
:---|:---:|:---:
New Django app and URLs for issuing reports. | <ul><li>- [x] </li></ul> | <ul><li>- [ ] </li></ul> |
Custom links for users in the Admin panel to view their reports. | <ul><li>- [x] </li></ul> | <ul><li>- [ ] </li></ul> |
Filtering data for an organization, and for a particular month (URL only in this task). | <ul><li>- [x] </li></ul> | <ul><li>- [ ] </li></ul> 
Generate invoices, by computing the number of days to bill. | <ul><li>- [x] </li></ul> | <ul><li>- [ ] </li></ul> 
Show only relevant AppServers (The ones that were active in the period). | <ul><li>- [x] </li></ul> | <ul><li>- [ ] </li></ul>
Simple HTML report, that can be copy+pasted into invoices. | <ul><li>- [x] </li></ul> | <ul><li>- [ ] </li></ul>

### Test instructions
- Create an Organization
- Create OpenEdxInstance
- Create WatchedFork and link it to that organization.
- Create Watched PR and link them with WatchedFork and the created instance.
- Spawn new AppServers
- Go to the [Organizations](http://localhost:5000/admin/userprofile/organization/) page in the admin panel
- Click on [Invoice Report]() link that will take you to the Invoice page.
- Repeat the process above to cover all test casses (AppServer that's terminated in the previous billing month, AppServer that's still running the current billing month, AppServer that's created after the billing month, ..)
- Check that the number of computed working days for each AppServer are correct.
- Check the number of computed charge per each AppServer, and fork, and all forks are computed correctly.
- Try copying and pasting the table in different platforms and check whether it keeps the correct format after the paste.
- Try changing the month and the year from the url and check the above again.

### Screenshots
<img width="1698" alt="screen shot 2018-09-30 at 3 29 52 am" src="https://user-images.githubusercontent.com/11036472/46256562-3ed7c500-c461-11e8-8d76-49872b1eec77.png">
<img width="1698" alt="screen shot 2018-09-30 at 3 29 47 am" src="https://user-images.githubusercontent.com/11036472/46256561-3d0e0180-c461-11e8-8477-2351055c7140.png">
<img width="1298" alt="screen shot 2018-10-01 at 4 37 47 am" src="https://user-images.githubusercontent.com/11036472/46286367-ca218b00-c533-11e8-925d-d4e04c785f3b.png">
